morebits.js

```diff
@@ -4566,215 +4577,1725 @@ Morebits.wiki.page = function(pageName, status) {
+	/**
+	 * @param {string|string[]} removeGroups - String or array of user group(s)
+	 * Forgiving: anything invalid is simply ignored by the API with a warning.
+	 */
+	this.setRemoveGroups = function(removeGroups) {
+		ctx.removeGroups = removeGroups;
+	};
+
+	/* Notification setters */
+	/** @param {boolean} [notifyBots=false] */
+	this.setNotifyBots = function(notifyBots) {
+		ctx.notifyBots = !!notifyBots;
+	};
+	/** @param {boolean} [notifyIndef=false] - Whether to notify users who are indefinitely blocked sitewide. */
+	this.setNotifyIndef = function(notifyIndef) {
+		ctx.notifyIndef = !!notifyIndef;
+	};
+	/** @param {boolean} [notifySelf=false] */
+	this.setNotifySelf = function(notifySelf) {
+		ctx.notifySelf = !!notifySelf;
+	};
+	/**
+	 * Provide templates and/or an external link, any of which, if
+	 * detected, will result in skipping a talkpage notification.  Can be
+	 * provided before or after the user is loaded.
+	 *
+	 * @param {string} [link] - An external link, either `http`, `https`,
+	 * or with no protocol provided.
+	 * @param {string|string[]} [templates] - A template or array of
+	 * templates; must include the namespace.
+	 */
+	this.setNotifySkips = function(link, templates) {
+		if (link) {
+			// Remove leading protocol, be http/https insensitive
+			ctx.notifySkipLink = link.replace(/^https?:\/\//, '');
+		}
+		if (templates) {
+			if (!Array.isArray(templates)) {
+				templates = [templates];
+			}
+			// The API will kindly ignore underscores, but if we set this
+			// before loading the page, we'll need to be able to compare
+			// the results to this list.  Alternativey, we could do regex
+			// matching in fnProcessNotify rather than checking for dups.
+			ctx.notifySkipTemplates = Morebits.array.uniq(templates).map(function(template) {
+				return template.replace(/_/, ' ');
+			});
+		}
+	};
+	/**
+	 * Set the text of the notification to be appended to user's talk
+	 * page.  If `setSectionTitle` is not used, should also contain
+	 * wikitext for the section title.
+	 *
+	 * @param {string} message
+	 */
+	this.setMessage = function(message) {
+		ctx.message = message;
+	};
+	/**
+	 * Create a new section, using this as the section title.
+	 * `setMessage` will set the section body.
+	 *
+	 * @param {string} title
+	 */
+	this.setSectionTitle = function(title) {
+		ctx.sectionTitle = title;
+	};
+	/**
+	 * Define an object of functions and values to apply to the
+	 * Morebits.wiki.page object used to notify the user talk page in
+	 * question.  Will be performed last, so is useful for applying
+	 * additional bespoke parameters, such as `setMinorEdit` or more complex
+	 * watch options to `setWatchlist`.
+	 *
+	 * @param {object} pageobjectFunctions - An object with `{function:
+	 * functionValue}` parameters.  Each key is the name of a
+	 * {@link Morebits.wiki.page} function, and its value is what will be
+	 * provided to that function.
+	 */
+	this.setPageobjectFunctions = function(pageobjectFunctions) {
```

Comment:
This is hacky.  Basically, we want to be able to affect the `Morebits.wiki.page` object used to send the notification, but with minimal duplication of efforts.  What happens if `setPageobjFunctions` provides something like `append`?  It allows for additional things like `setMinor`, etc., but is weird.


It's required for full handling of the suite of `watchlistEnums`, which at the moment is xfd, speedy, and image.  That's maybe enough to justify doing something more involved within `Morebits.wiki.user` but that has its own pitfalls, since `watchuser` is binary.

Reply:
Is there a reason not to just expose the underlying Morebits.wiki.page object?